File: Leetcode/Trapping Rainwater Problem/Trapping_Rainwater.py
#Question Link <https://leetcode.com/problems/trapping-rain-water/>
class Solution:
    def trap(self, height):
        left, right = 0, len(height) - 1
        ans = left_max = right_max = 0
        while left < right:
            if height[left] < height[right]:
                left_max = max(left_max, height[left])
                ans += left_max - height[left]
                left += 1
            else:
                right_max = max(right_max, height[right])
                ans += right_max - height[right]
                right -= 1
        return ans

File: Leetcode/Trapping rain water/trapping_rain_water.cpp
// Problem link - https://leetcode.com/problems/trapping-rain-water/

class Solution
{
public:
    int trap(vector<int> &height)
    {
        int res = 0;
        int sz = height.size(), left = 0, right = sz - 1;
        int leftMax = 0, rightMax = 0;

        while (left <= right)
        {
            if (height[left] <= height[right])
            {
                if (leftMax < height[left])
                {
                    leftMax = height[left];
                }
                else
                {
                    res += leftMax - height[left];
                    ++left;
                }
            }
            else
            {
                if (height[right] > rightMax)
                {
                    rightMax = height[right];
                }
                else
                {
                    res += rightMax - height[right];
                    --right;
                }
            }
        }

        return res;
    }
};